### PR TITLE
Multichain Identity Indexer

### PR DESCRIPTION
**Pull request by vikiival**
#### Project Abstract

The Multichain Identity Indexer - Identics is a GraphQL service designed to streamline the interaction with Identity pallet that was originally on Relay-chain and now resides on People Chain. This service acts as a robust data layer for socially-oriented dApps, leveraging the simplicity of GraphQL and REST API. The indexer addresses the challenge of querying identity information for multiple accounts in parallel and bridges the gap between historical data on the Relay chain and real-time updates on People Chain.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (USDC & DOT) address in the application and bank details via email, if applicable).
- [x] I understand that an agreed upon percentage of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by semuelle — CHANGES_REQUESTED** [review]
Thanks for the application, @vikiival. Other than the use case you have been working on, I'm wondering who or what would require indexing of the history of identities?

File: applications/multichain_identity_indexer.md
Comment (semuelle) [inline applications/multichain_identity_indexer.md:145]: I get a `403 Forbidden` when trying to access this file.

**Comment by vikiival** [comment]
> Other than the use case you have been working on, I'm wondering who or what would require indexing of the history of identities?

Hey, 
that is a very good question!
I have split it into several topics that could leverage the identity history

1. Governance Analytics & Reputation Systems

- Tracking how long accounts have maintained verified identities before participating in important governance votes (any opengov platform)
- Building reputation systems that consider the longevity and stability to identify trusted community members 

2. Security & Fraud Prevention

- Detecting suspicious patterns of identity changes or clearances that might indicate malicious behavior
- Helping registrars review an account's identity history before providing judgements
- Enabling apps to flag accounts that frequently change or clear their identities 

3. Social Network Analysis

- Understanding the evolution of relationships between accounts (social graph)
- Analyzing the growth of identities within the ecosystem

4. Economic Research & Analysis

- Studying how identity verification impacts user behavior and economic activity
- Understanding how identity systems are used across different market conditions

**Review by semuelle — APPROVED** [review]
Hi @vikiival, thanks for the thorough reply and sorry for the long wait. Some of the use cases and future plans you listed seem very niche to me, but looking at ENS, there is still a lot of room for improvement regarding identities in our ecosystem, so I am happy to approve your application and share it with the rest of the committee.

**Review by Noc2 — DISMISSED** [review]
It would be nice to get open-source software like this potentially funded via the treasury (see for example [Open Source Developer Grants Bounty Program](https://polkadot.polkassembly.io/referenda/1080)), given that the treasury is the long-term funding solution for these things. But I'm happy to also support it here.

File: applications/multichain_identity_indexer.md
Comment (takahser) [inline applications/multichain_identity_indexer.md:231]:
```suggestion
- **Referrer:** Luu (ex Polkadot Ambassador / Meetups Bounty Curator)
```

**Review by takahser — CHANGES_REQUESTED** [review]
@vikiival This looks very interesting.

However, I have a couple of questions:

- Will it be possible to query a complete list of available identities, along with additional metadata like their age or registration timestamps? This could be valuable for features like airdrops.
- Once developed, do you plan to host the indexer as a public-facing API for broader community use? This could be especially useful for developers looking to build dApps without setting up their infrastructure.
- Will this indexer simplify running registrars? If that’s a popular use case, it might make sense to add a demo project showcasing all the API calls needed for registrar-related extrinsics. This could help onboard users faster.

**Comment by vikiival** [comment]
> * Will it be possible to query a complete list of available identities, along with additional metadata like their age or registration timestamps? This could be valuable for features like airdrops.

@takahser Amazing idea!
Yes, Identity entity will contain `createdAt: DateTime! @index` and `updatedAt: DateTime! @index` along with historical changes. 
It is a perfect use-case for loyal Polkadot / Kusama users.

> * Once developed, do you plan to host the indexer as a public-facing API for broader community use? This could be especially useful for smaller/early-stage dApp dev teams, since they wouldn't have to set up their own infrastructure.

We plan to host Identics on SubSquid infra and Hono API as a public good solution. I agree that identity management takes a lot of precious dev time.


> * Will this indexer simplify running registrars? If that's a popular use case, it might make sense to add a demo project showcasing all the API calls needed for registrar-related extrinsic.

That's a very good and valid question. As far as I know, currently, registrars have to listen to the RPC events and check for judgement requests. Additionally, they need to parse the data from the RPC node (such as mail, Twitter, etc.) to reach the user. 
Identics simplify this as all registrars must do is `subscribe` to the GraphQL server. Changes will be pushed to the subscribed registrars as soon as a block is finalised.  

We will also add WIKI with queries/examples that can be used. 

Misc
---

Can I ask @semuelle and @Noc2 for approval again? Seems that force-push dismissed the approval

**Review by semuelle — DISMISSED** [review]

**Review by takahser — CHANGES_REQUESTED** [review]
@vikiival thanks for your prompt response.

> We will also add WIKI with queries/examples that can be used.

Could you add this to the proposal? After that I'd be happy to add my approval as well. 

Btw, you don't have to worry about the dismissal of approvals - as long as you don't remove things reviewers would usually just re-approve.

**Comment by vikiival** [comment]
> Could you add this to the proposal? After that I'd be happy to add my approval as well.

Added as delivery `0e` (Line 170)
 
> Btw, you don't have to worry about the dismissal of approvals - as long as you don't remove things reviewers would usually just re-approve.

Oh, I did not know ^-^.

**Review by takahser — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Review by semuelle — APPROVED** [review]

**Comment by keeganquigley** [comment]
Hi @vikiival it looks like M1 was due about 3 months ago, can you provide a progress update? Thanks.

**Comment by vikiival** [comment]
Hey @keeganquigley, 
Thanks for the ping. 
I am aware of the delivery delay, which is primarily due to Adam's time constraints (since he is on Erasmus and has less time than expected)

Project updates

1. Indexer is correctly processing events from Polkadot
Screenshot 2025-05-09 at 16 22 25
2. Indexer is correctly processing events from Polkadot People
3. Created schema to store data

Missing:
1. API
2. Better schema design
3. System to optimally handle multiple chains into one "entity"

We will dedicate more capacity to this project to successfully deliver the M1.

**Comment by keeganquigley** [comment]
Hi @vikiival hope your summer is going well. Any updates here?

**Comment by vikiival** [comment]
> Hi @vikiival hope your summer is going well. Any updates here?

Hey @keeganquigley, 
we have successfully implemented multi-support for `Polkadot` relaychain, therefore we are very close to the delivery.

Alongside we are up-to date with current chain-specs - there was an update on People chain

We apologize for the delay 🙏 🥺 

Screenshot 2025-08-12 at 23 07 33

Alongside we have implemented squid-native-REST-API, this was kinda painful process as we need to dive into the subsquid internals

Screenshot 2025-08-13 at 00 55 22

**Pull request merged**
